lib/core/include/scipp/core/dict.h

```diff
@@ -0,0 +1,423 @@
+    m_keys.reserve(new_capacity);
+    m_values.reserve(new_capacity);
+  }
+
+  [[nodiscard]] bool contains(const Key &key) const noexcept {
+    return find_key(key) != m_keys.end();
+  }
+
+  template <class V> void insert_or_assign(const key_type &key, V &&value) {
+    if (const auto key_it = find_key(key); key_it == m_keys.end()) {
+      m_keys.push_back(key);
+      m_values.emplace_back(std::forward<V>(value));
+    } else {
+      m_values[index_of(key_it)] = std::forward<V>(value);
+    }
+  }
+
+  void erase(const key_type &key) { static_cast<void>(extract(key)); }
+
+  mapped_type extract(const key_type &key) {
+    const auto key_it = expect_find_key(key);
+    m_keys.erase(key_it);
+    const auto value_it = std::next(m_values.begin(), index_of(key_it));
+    mapped_type value = std::move(*value_it);
+    m_values.erase(value_it);
+    return value;
+  }
+
+  void clear() {
+    m_keys.clear();
+    m_values.clear();
+  }
+
+  [[nodiscard]] const mapped_type &operator[](const key_type &key) const {
+    return m_values[expect_find_index(key)];
+  }
+
+  [[nodiscard]] mapped_type &operator[](const key_type &key) {
+    return m_values[expect_find_index(key)];
+  }
+
+  [[nodiscard]] const mapped_type &at(const key_type &key) const {
+    return (*this)[key];
+  }
+
+  [[nodiscard]] mapped_type &at(const key_type &key) { return (*this)[key]; }
+
+  [[nodiscard]] const_iterator find(const key_type &key) const {
+    if (const auto key_it = find_key(key); key_it == m_keys.end()) {
+      return end();
+    } else {
+      return const_iterator(m_keys, key_it,
+                            std::next(m_values.begin(), index_of(key_it)));
+    }
+  }
+
+  [[nodiscard]] iterator find(const key_type &key) {
+    if (const auto key_it = find_key(key); key_it == m_keys.end()) {
+      return end();
+    } else {
+      return iterator(m_keys, key_it,
+                      std::next(m_values.begin(), index_of(key_it)));
+    }
+  }
+
+  [[nodiscard]] auto keys_begin() const noexcept {
+    return const_key_iterator(m_keys, m_keys.cbegin());
+  }
+
+  [[nodiscard]] auto keys_end() const noexcept {
+    return const_key_iterator(m_keys, m_keys.cend());
+  }
+
+  [[nodiscard]] auto values_begin() noexcept {
+    return value_iterator(m_values, m_values.begin());
+  }
+
+  [[nodiscard]] auto values_end() noexcept {
+    return value_iterator(m_values, m_values.end());
+  }
+
+  [[nodiscard]] auto values_begin() const noexcept {
+    return const_value_iterator(m_values, m_values.begin());
+  }
+
+  [[nodiscard]] auto values_end() const noexcept {
+    return const_value_iterator(m_values, m_values.end());
+  }
+
+  [[nodiscard]] auto begin() noexcept {
+    return iterator(m_keys, m_keys.cbegin(), m_values.begin());
+  }
+
+  [[nodiscard]] auto end() noexcept {
+    return iterator(m_keys, m_keys.cend(), m_values.end());
+  }
+
+  [[nodiscard]] auto begin() const noexcept {
+    return const_iterator(m_keys, m_keys.cbegin(), m_values.begin());
```

Comment:
```suggestion
    return const_iterator(m_keys, m_keys.cbegin(), m_values.cbegin());
```
? Probably the same since `const_iterator` is used, but helps readability. Same below and above.

Reply:
Below, yes. Above, no. The non-const `begin` and `end` need to return `m_keys.cbegin()` but `m_values.begin()` (same with end).